### PR TITLE
Allow dict input_spec for keras.Model

### LINKED ISSUE #42691: Cannot build network with a dict input_spec
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04 (in docker container)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.2.0-rc4-8-g2b96f3662b 2.2.0
- Python version: 3.6.9
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: 10.2, N/A
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**

Cannot build a model with a `dict` input_spec (`model.build`)

**Describe the expected behavior**

The model should be able to build the graph with a `dict` input_spec:

1. We can call a model with dictionary inputs, so specifying that spec as dictionary seems natural
2. I believe this is what the developers intended too, given [this line](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/engine/network.py#L650) in the current master branch. 

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

```python
In [33]: class TestModel(keras.Model):
    ...:     def __init__(self, *args, **kwargs):
    ...:         super(TestModel, self).__init__(*args, **kwargs)
    ...:         self.layer = keras.layers.Dense(10)
    ...:     def call(self, inputs):
    ...:         return {'b': self.layer(inputs['a'])}
    ...: 

In [34]: model = TestModel()

In [35]: model.build({ 'a': (None, 5) })
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-35-f5e425698cd1> in <module>
----> 1 model.build({ 'a': (None, 5) })

/usr/local/lib/python3.6/dist-packages/tensorflow/python/keras/engine/network.py in build(self, input_shape)
    633                        'Please specify a batch input shape of type tuple or '
    634                        'list of input shapes. User provided '
--> 635                        'input type: {}'.format(type(input_shape)))
    636
    637     if input_shape and not self.inputs:

ValueError: Specified input shape is not one of the valid types. Please specify a batch input shape of type tuple or list of input shapes. User provided input type: <class 'dict'>

In [36]: model({'a': np.random.random((3,5)).astype(np.float32)})
Out[36]: 
{'b': <tf.Tensor: shape=(3, 10), dtype=float32, numpy=
 array([[ 0.57159793, -0.54181933, -0.05867613, -0.13133162,  0.33426586,
          0.88725936,  0.6102787 ,  0.25660542, -0.32508045,  0.18882477],
        [ 0.5146243 , -0.07790124, -0.13295858,  0.33439606,  0.6062204 ,
          0.67318   ,  0.7965131 , -0.09740189, -0.49425927, -0.3331495 ],
        [ 0.828179  , -0.3307852 , -0.49201646, -0.16802055,  0.44089833,
          0.8766354 ,  0.92791915,  0.28562796, -0.8893677 , -0.428018  ]],
       dtype=float32)>}
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

### PR DESCRIPTION
**Pull request by yongtang**
This PR tries to address the issue raised in #42691 where
usage of dict will cause ValueError in keras.Model.

This PR fixes #42691.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/python/keras/engine/training.py [hunk 1]
```diff
@@ -350,7 +350,7 @@ def build(self, input_shape):
     if input_shape is None:
       raise ValueError('Input shape must be defined when calling build on a '
                        'model subclass network.')
-    valid_types = (tuple, list, tensor_shape.TensorShape)
+    valid_types = (tuple, list, dict, tensor_shape.TensorShape)
     if not isinstance(input_shape, valid_types):
       raise ValueError('Specified input shape is not one of the valid types. '
                        'Please specify a batch input shape of type tuple or '
```
Comment (qlzh727) [inline tensorflow/python/keras/engine/training.py:356]: Please add 'dict' here in the error message.

**Review by qlzh727 — DISMISSED** [review]

**Comment by yongtang** [comment]
Thanks @qlzh727 for the review. The PR has been updated. Please take a look.

**Comment by qlzh727** [comment]
> Thanks @qlzh727 for the review. The PR has been updated. Please take a look.

Somehow the test file is removed from the PR. Can u please add it back? Thanks.

**Comment by gbaned** [comment]
@yongtang Can you please check @qlzh727's comments and keep us posted ? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
@yongtang Any update on this PR? and please resolve conflicts Thanks!

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!

**Pull request closed without merging**